src/votekit/ballot_generator.py

```
@@ -1201,4 +1201,144 @@ def __init__(self, **data):
+        blocs = list(self.bloc_voter_prop.keys())
+        bloc_props = list(self.bloc_voter_prop.values())
+        ballots_per_block = dict(
+            zip(blocs, apportion.compute("huntington", bloc_props, number_of_ballots))
+        )
+
+        pref_profile_by_bloc = {}
+
+        for bloc in blocs: 
+            ballot_pool = []
+            # number of voters in this bloc
+            num_ballots = ballots_per_block[bloc]
+
+            # TODO the code below assumes only two blocs
+            ballot_type_distribution = [self.cohesion_parameters[bloc], 
+                                        1-self.cohesion_parameters[bloc]]
+            non_zero_cands = [c for c,v in self.pref_interval_by_bloc[bloc].items() if v>0]
+
+            opp_bloc = set(blocs).difference(bloc).pop()
+            
+            for _ in range(num_ballots):
+                # sample ballot type
+                ballot_type = []
+                candidate_count = {b:0 for b in blocs}
+                for _ in range(len(non_zero_cands)):
+                    # choose a bloc as the next slot in the ballot
+                    cand_bloc = list(np.random.choice([bloc, opp_bloc], size =1, 
+                                                      p = ballot_type_distribution))[0]
+                    ballot_type.append(cand_bloc)
+                    candidate_count[cand_bloc] += 1
+
+                    # if there are more candidates to fill
+                    if len(ballot_type) < len(non_zero_cands):
+                        for b in blocs:
+                            # and you are out of non-zero candidates for a slate
+                            if candidate_count[b] == len([c for c in self.slate_to_candidates[b] 
+                                                        if self.pref_interval_by_bloc[bloc][c]>0]):
+                                 # TODO this assumes only two blocs
+                                 o_bloc = set(blocs).difference(b).pop()
```

same note as above regarding failing loudly if there are multiple blocs